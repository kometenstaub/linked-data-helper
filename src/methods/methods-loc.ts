import { App, FileSystemAdapter, normalizePath, Notice } from 'obsidian';
import type LinkeDataHelperPlugin from '../main';
import type { Graph, headings, LcshInterface } from '../interfaces';
import { createReadStream, writeFileSync } from 'fs';
import { normalize } from 'path';
import split2 from 'split2';

export class SkosMethods {
    app: App;
    plugin: LinkeDataHelperPlugin;

    constructor(app: App, plugin: LinkeDataHelperPlugin) {
        this.app = app;
        this.plugin = plugin;
    }

    async getAbsolutePath(fileName: string): Promise<string> {
        let basePath;
        //let relativePath;
        // base path
        if (this.app.vault.adapter instanceof FileSystemAdapter) {
            basePath = this.app.vault.adapter.getBasePath();
        } else {
            throw new Error('Cannot determine base path.');
        }
        // relative path
        //relativePath = `${this.app.vault.configDir}/plugins/linked-data-vocabularies/${fileName}`;
        let attachmentPath =
            await this.app.vault.getAvailablePathForAttachments(
                fileName,
                'json'
            );
        // overwrite file
        attachmentPath = attachmentPath.replace(/(.+?)(?: 1)(\.json)/, '$1$2');
        // absolute path
        return normalize(`${basePath}/${attachmentPath}`);
    }

    public convertLcshSkosNdjson(outputPath?: string) {
        let jsonPrefLabel: headings[] = [];
        let subdivisions: headings[] = [];
        let jsonUriToPrefLabel = {};
        let inputPath = '';
        if (this.plugin.settings.lcshInputPath) {
            inputPath = this.plugin.settings.lcshInputPath;
        } else {
            const text =
                'Please specify an input path for LCSH in the settings.';
            new Notice(text);
            throw Error(text);
        }
        let newOutputPath = '';
        if (outputPath) {
            newOutputPath = outputPath;
        }
        createReadStream(inputPath)
            .pipe(split2(JSON.parse))
            .on('data', (obj: LcshInterface) => {
                //@ts-ignore
                let currentObj: headings = {};
                const id = obj['@context'].about;
                for (let element of obj['@graph']) {
                    let broaderURLs: string[] = [];
                    let narrowerURLs: string[] = [];
                    let relatedURLs: string[] = [];
                    let prefLabel: string = '';
                    let altLabel: string = '';

                    //if (
                    //    element['skos:inScheme']?.['@id'] ===
                    //    'http://id.loc.gov/authorities/subjects'
                    //) {
                    let uri = '';
                    if (element['skos:prefLabel']?.['@language'] === 'en') {
                        uri = element['@id'];
                        const currentPrefLabel =
                            element['skos:prefLabel']['@value'];
                        if (uri === id) {
                            prefLabel = currentPrefLabel;
                        } else {
                            continue;
                        }
                    } else {
                        continue;
                    }
                    //@ts-expect-error
                    const endUri: string = uri.split('/').last();
                    Object.assign(jsonUriToPrefLabel, {
                        [endUri]: prefLabel,
                    });
                    if (element['skos:altLabel']?.['@language'] === 'en') {
                        altLabel = element['skos:altLabel']['@value'];
                    }
                    broaderURLs = this.pushHeadings(element, 'broader');
                    narrowerURLs = this.pushHeadings(element, 'narrower');
                    relatedURLs = this.pushHeadings(element, 'related');
                    currentObj = this.onlyReturnFull(
                        prefLabel,
                        altLabel,
                        uri,
                        broaderURLs,
                        narrowerURLs,
                        relatedURLs
                    );
                    if (element['skos:note']) {
                        let note = element['skos:note'];
                        if (Array.isArray(note)) {
                            let newNote = '';
                            for (let el of note) {
                                newNote += el;
                            }
                            note = newNote;
                        }
                        currentObj.note = note;
                        if (note.includes('Use as a')) {
                            subdivisions.push(currentObj);
                        } else {
                            jsonPrefLabel.push(currentObj);
                        }
                        //} else if (element['skos:editorial']) {
                        //  let editorial = element['skos:editorial'];
                        //  if (Array.isArray(editorial)) {
                        //      let neweditorial = '';
                        //      for (let el of editorial) {
                        //          neweditorial += el;
                        //      }
                        //      editorial = neweditorial;
                        //  }
                        //  currentObj.note = editorial;
                        //  if (editorial.startsWith('subdivision')) {
                        //      subdivisions.push(currentObj)
                        //  } else {
                        //      jsonPrefLabel.push(currentObj)
                        //  }
                    } else {
                        jsonPrefLabel.push(currentObj);
                    }

                    //Object.assign(jsonPrefLabel, currentObj)
                    break;
                    //}
                }
            })
            .on('end', () => {
                let jsonPrefPath = '';
                let jsonUriPath = '';
                let jsonSubdivPath = '';
                const { adapter } = this.app.vault;
                if (newOutputPath === '') {
                    const attachmentFolder = normalizePath(
                        this.app.vault.config.attachmentFolderPath +
                            '/' +
                            'linked-data-vocabularies/'
                    );
                    (async () => {
                        const isDir = await adapter.exists(attachmentFolder);
                        if (!isDir) {
                            adapter.mkdir(attachmentFolder);
                        }
                    })();
                    // prettier-ignore
                    (async () => {
                        jsonPrefPath = await this.getAbsolutePath(
                            'linked-data-vocabularies/lcshSuggester'
                        );
                        jsonUriPath = await this.getAbsolutePath(
                            'linked-data-vocabularies/lcshUriToPrefLabel'
                        );
                        jsonSubdivPath = await this.getAbsolutePath(
                            'linked-data-vocabularies/lcshSubdivSuggester'
                      );
                        writeFileSync(jsonPrefPath, JSON.stringify(jsonPrefLabel));
                        writeFileSync(jsonUriPath, JSON.stringify(jsonUriToPrefLabel));
                        writeFileSync(jsonSubdivPath, JSON.stringify(subdivisions));
                    })();
                } else {
                    jsonPrefPath = normalize(
                        newOutputPath + '/' + 'lcshSuggester.json'
                    );
                    jsonUriPath = normalize(
                        newOutputPath + '/' + 'lcshUriToPrefLabel.json'
                    );
                    jsonSubdivPath = normalize(
                        newOutputPath + '/' + 'lcshSubdivSuggester.json'
                    );
                    writeFileSync(jsonPrefPath, JSON.stringify(jsonPrefLabel));
                    // prettier-ignore
                    writeFileSync(jsonUriPath, JSON.stringify(jsonUriToPrefLabel));
                    writeFileSync(jsonSubdivPath, JSON.stringify(subdivisions));
                }

                new Notice('The three JSON files have been written.');
            });
    }

    private onlyReturnFull(
        prefLabel: string,
        altLabel: string,
        uri: string,
        broaderURLs: string[],
        narrowerURLs: string[],
        relatedURLs: string[]
    ): headings {
        //@ts-ignore
        let currentObj: headings = {};
        let reducedBroaderURLs: string[] = [];
        for (let url of broaderURLs) {
            //@ts-expect-error
            reducedBroaderURLs.push(url.split('/').last());
        }
        let reducedNarrowerURLs: string[] = [];
        for (let url of narrowerURLs) {
            //@ts-expect-error
            reducedNarrowerURLs.push(url.split('/').last());
        }
        let reducedRelatedURLs: string[] = [];
        for (let url of relatedURLs) {
            //@ts-expect-error
            reducedRelatedURLs.push(url.split('/').last());
        }

        //@ts-expect-error
        let reducedUri: string = uri.split('/').last();
        if (altLabel !== '') {
            if (broaderURLs.length > 0) {
                if (narrowerURLs.length > 0 && relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            bt: reducedBroaderURLs,
                            nt: reducedNarrowerURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else if (narrowerURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            bt: reducedBroaderURLs,
                            nt: reducedNarrowerURLs,
                        };
                    }
                } else if (relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            bt: reducedBroaderURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            bt: reducedBroaderURLs,
                        };
                    }
                }
            } else if (narrowerURLs.length > 0) {
                if (relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            nt: reducedNarrowerURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            aL: altLabel,
                            nt: reducedNarrowerURLs,
                        };
                    }
                }
            } else if (relatedURLs.length > 0) {
                currentObj = {
                    pL: prefLabel,
                    uri: reducedUri,
                    aL: altLabel,
                    rt: reducedRelatedURLs,
                };
            } else {
                currentObj = {
                    pL: prefLabel,
                    uri: reducedUri,
                    aL: altLabel,
                };
            }
        } else {
            if (broaderURLs.length > 0) {
                if (narrowerURLs.length > 0 && relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            bt: reducedBroaderURLs,
                            nt: reducedNarrowerURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else if (narrowerURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            bt: reducedBroaderURLs,
                            nt: reducedNarrowerURLs,
                        };
                    }
                } else if (relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            bt: reducedBroaderURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            bt: reducedBroaderURLs,
                        };
                    }
                }
            } else if (narrowerURLs.length > 0) {
                if (relatedURLs.length > 0) {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            nt: reducedNarrowerURLs,
                            rt: reducedRelatedURLs,
                        };
                    }
                } else {
                    {
                        currentObj = {
                            pL: prefLabel,
                            uri: reducedUri,
                            nt: reducedNarrowerURLs,
                        };
                    }
                }
            } else if (relatedURLs.length > 0) {
                currentObj = {
                    pL: prefLabel,
                    uri: reducedUri,
                    rt: reducedRelatedURLs,
                };
            } else {
                currentObj = {
                    pL: prefLabel,
                    uri: reducedUri,
                };
            }
        }
        return currentObj;
    }

    private pushHeadings(element: Graph, type: string): string[] {
        let urls = [];
        const headingType: string = `skos:${type}`;
        //@ts-ignore
        if (element[headingType]) {
            //@ts-ignore
            if (Array.isArray(element[headingType])) {
                //@ts-ignore
                for (let id of element[headingType]) {
                    urls.push(id['@id']);
                }
            } else {
                //@ts-ignore
                urls.push(element[headingType]['@id']);
            }
        }
        return urls;
    }
}
