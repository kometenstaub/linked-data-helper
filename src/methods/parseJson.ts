import type {Graph, headings, LcshInterface, SkolemGraphNode, uriToHeading,} from '../interfaces';


export function parseJsonHeading(
    obj: LcshInterface,
    jsonPrefLabel: headings[],
    subdivisions: headings[],
    jsonUriToPrefLabel: uriToHeading
) {
    //@ts-expect-error, it needs to be initialised
    // and will be populated later on
    let currentObj: headings = {};
    const id = obj['@context'].about;
    for (const element of obj['@graph']) {
        let broaderURLs: string[] = [];
        let narrowerURLs: string[] = [];
        let relatedURLs: string[] = [];
        let prefLabel = '';
        let altLabel = '';
        let lcc = '';

        let uri = '';
        if (element['skos:prefLabel']?.['@language'] === 'en') {
            uri = element['@id'];
            const currentPrefLabel = element['skos:prefLabel']['@value'];
            if (uri === id) {
                prefLabel = currentPrefLabel;
            } else {
                continue;
            }
        } else {
            continue;
        }
        if (element['madsrdf:classification']) {
            const skolemIri: string = element['madsrdf:classification']['@id'];
            for (const againElement of obj['@graph']) {
                if (againElement['@id'] === skolemIri) {
                    const skolemNode = againElement as SkolemGraphNode;
                    lcc = skolemNode['madsrdf:code'];
                    break;
                }
            }
        }
        const splitUri: string[] = uri.split('/');
        const endUri = splitUri[splitUri.length - 1];
        Object.assign(jsonUriToPrefLabel, {
            [endUri]: prefLabel,
        });
        if (element['skos:altLabel']?.['@language'] === 'en') {
            altLabel = element['skos:altLabel']['@value'];
        }
        const graph = obj['@graph'];
        broaderURLs = pushHeadings(element, graph, 'broader');
        narrowerURLs = pushHeadings(element, graph, 'narrower');
        relatedURLs = pushHeadings(element, graph, 'related');
        currentObj = onlyReturnFull(
            prefLabel,
            altLabel,
            uri,
            broaderURLs,
            narrowerURLs,
            relatedURLs,
            lcc
        );
        if (element['skos:note']) {
            let note = element['skos:note'];
            if (Array.isArray(note)) {
                let newNote = '';
                for (const el of note) {
                    newNote += el;
                }
                note = newNote;
            }
            currentObj.note = note;
            if (note.includes('Use as a')) {
                subdivisions.push(currentObj);
            } else {
                jsonPrefLabel.push(currentObj);
            }
            //} else if (element['skos:editorial']) {
            //  let editorial = element['skos:editorial'];
            //  if (Array.isArray(editorial)) {
            //      let neweditorial = '';
            //      for (let el of editorial) {
            //          neweditorial += el;
            //      }
            //      editorial = neweditorial;
            //  }
            //  currentObj.note = editorial;
            //  if (editorial.startsWith('subdivision')) {
            //      subdivisions.push(currentObj)
            //  } else {
            //      jsonPrefLabel.push(currentObj)
            //  }
        } else {
            jsonPrefLabel.push(currentObj);
        }
        break;
    }
}

function splitUri(url: string): string {
    const splitUrl = url.split('/');
    return splitUrl[splitUrl.length - 1]
}


function onlyReturnFull(
    prefLabel: string,
    altLabel: string,
    uri: string,
    broaderURLs: string[],
    narrowerURLs: string[],
    relatedURLs: string[],
    lcc: string
): headings {
    //@ts-expect-error, the object needs to be initialised,
    // it is populated later on
    const currentObj: headings = {};
    const reducedBroaderURLs: string[] = [];
    for (const url of broaderURLs) {
        if (url && url.includes('/')) {
            const endUri = splitUri(url)
            reducedBroaderURLs.push(endUri);
        } else {
            reducedBroaderURLs.push(url);
        }
    }
    const reducedNarrowerURLs: string[] = [];
    for (const url of narrowerURLs) {
        if (url && url.includes('/')) {
            const endUri = splitUri(url)
            reducedNarrowerURLs.push(endUri);
        } else {
            reducedNarrowerURLs.push(url);
        }
    }
    const reducedRelatedURLs: string[] = [];
    for (const url of relatedURLs) {
        if (url && url.includes('/')) {
            const endUri = splitUri(url)
            reducedRelatedURLs.push(endUri);
        } else {
            reducedRelatedURLs.push(url);
        }
    }

    const reducedUri: string = splitUri(uri)
    currentObj.pL = prefLabel;
    currentObj.uri = reducedUri;
    if (altLabel !== '') {
        currentObj.aL = altLabel;
    }
    if (broaderURLs.length > 0) {
        currentObj.bt = reducedBroaderURLs;
    }
    if (narrowerURLs.length > 0) {
        currentObj.nt = reducedNarrowerURLs;
    }
    if (relatedURLs.length > 0) {
        currentObj.rt = reducedRelatedURLs;
    }
    if (lcc !== '') {
        currentObj.lcc = lcc;
    }

    return currentObj;
}

function pushHeadings(
    element: Graph,
    graph: Graph[],
    type: 'broader' | 'narrower' | 'related'
): string[] {
    const urls = [];
    const headingType:
        | 'skos:broader'
        | 'skos:narrower'
        | 'skos:related' = `skos:${type}`;
    const relation = element[headingType];
    if (relation !== undefined) {
        if (Array.isArray(relation)) {
            for (const subElement of relation) {
                const id = subElement['@id'];
                if (id.startsWith('_:')) {
                    const term = getSkolemIriRelation(graph, id);
                    urls.push(term);
                } else {
                    urls.push(id);
                }
            }
        } else {
            const id: string = relation['@id'];
            if (id.startsWith('_:')) {
                const term = getSkolemIriRelation(graph, id);
                urls.push(term);
            } else {
                urls.push(id);
            }
        }
    }

    return urls;
}

/**
 *
 * @param graph
 * @param id
 * @returns - it always returns a non-empty string, because this function is only called if there is a matching result
 */
function getSkolemIriRelation(graph: Graph[], id: string): string {
    let term = '';
    for (const part of graph) {
        if (part['@id'] === id) {
            const prefLabel = part['skos:prefLabel'];
            if (prefLabel['@language'] === 'en') {
                term = prefLabel['@value'];
                break;
            }
        }
    }
    return term;
}
