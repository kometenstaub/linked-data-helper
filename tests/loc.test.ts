import { readFileSync } from 'fs';
import { parseJsonHeading } from '../src/methods/parseJson';
import type {headings, LcshInterface, uriToHeading} from '../src/interfaces';
import {match} from "assert";

function readJson(fileName: string): LcshInterface {
    const currentFile = readFileSync(`tests/testFiles/${fileName}`, {
        encoding: 'utf-8',
    });
    return JSON.parse(currentFile);
}

function snapshotJsonParsing(filename: string, snapshot: boolean) {
    const obj = readJson(filename);
    const jsonPrefLabel: headings[] = [];
    const subdivisions: headings[] = [];
    const jsonUriToPrefLabel: uriToHeading= {};
    parseJsonHeading(obj, jsonPrefLabel, subdivisions, jsonUriToPrefLabel);
    if (snapshot) {
        matchSnapshot(jsonPrefLabel, subdivisions, jsonUriToPrefLabel)
    } else {
        return {jsonPrefLabel, subdivisions, jsonUriToPrefLabel }
    }
}

function matchSnapshot(jsonPrefLabel: headings[], subdivisions: headings[], jsonUriToPrefLabel: uriToHeading) {
    expect(jsonPrefLabel).toMatchSnapshot();
    expect(subdivisions).toMatchSnapshot();
    expect(jsonUriToPrefLabel).toMatchSnapshot();
}


/**
 * unit test
 */
test('History URI is correct', () => {
    const obj = readJson('History.json');
    expect(obj['@context'].about).toBe(
        'http://id.loc.gov/authorities/subjects/sh85061227'
    );
});

/**
 * Snapshot tests
 */

test('Snapshot of JSON parsing logic History.json', () => {
    snapshotJsonParsing('History.json', true);
});

test('Snapshot of JSON parsing logic History-subdiv.json', () => {
    snapshotJsonParsing('Archaeology.json', true);
});

test('Snapshot of JSON parsing logic Archaeology.json', () => {
    snapshotJsonParsing('Archaeology.json', true);
});

test('Snapshot of JSON parsing logic Obsidian.json', () => {
    snapshotJsonParsing('Obsidian.json', true);
});

/**
 * more unit tests
 */

//test('Archaeology properties of JSON objects', () => {
//    snapshotJsonParsing('Archaeology.json', false);
//});
