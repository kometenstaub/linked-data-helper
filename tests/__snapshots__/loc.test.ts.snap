// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of JSON parsing logic Archaeology.json 1`] = `
[
  {
    "aL": [
      "Archeology",
    ],
    "bt": [
      "sh85005581",
      "sh85010480",
      "sh85061212",
    ],
    "lcc": "CC",
    "note": "Here are entered works on archaeology as a branch of learning. This heading may be divided geographically for works on this branch of learning in a specific place. Works on the antiquities of particular regions, countries, cities, etc. are entered under the name of the place subdivided by [Antiquities.]",
    "nt": [
      "sh00007440",
      "sh2003003476",
      "sh2003004186",
      "sh2003009064",
      "sh2003009199",
      "sh2003011000",
      "sh2004001262",
      "sh2005001576",
      "sh2006000287",
      "sh2007006623",
      "sh2009002956",
      "sh2009004421",
      "sh2009004470",
      "sh2012000959",
      "sh2012004635",
      "sh2012004636",
      "sh2013000143",
      "sh2013001302",
      "sh2013001303",
      "sh2015002234",
      "sh2021005077",
      "sh2021006222",
      "sh85001256",
      "sh85004685",
      "sh85006703",
      "sh85018080",
      "sh85027024",
      "sh85034228",
      "sh85046105",
      "sh85048843",
      "sh85050977",
      "sh85053066",
      "sh85060817",
      "sh85061115",
      "sh85065024",
      "sh85065824",
      "sh85066643",
      "sh85072567",
      "sh85079786",
      "sh85087721",
      "sh85093636",
      "sh85097895",
      "sh85101308",
      "sh85102008",
      "sh85107780",
      "sh85109300",
      "sh85114683",
      "sh85116926",
      "sh85126368",
      "sh85126691",
      "sh85127907",
      "sh85134099",
      "sh85139611",
      "sh85144346",
      "sh86001404",
      "sh87000460",
      "sh89000928",
      "sh89002313",
      "sh91002758",
      "sh93006952",
      "sh94001190",
      "sh94008300",
      "sh95000886",
      "sh95000887",
    ],
    "pL": "Archaeology",
    "rt": [
      "sh85005757",
    ],
    "uri": "sh85006507",
  },
]
`;

exports[`Snapshot of JSON parsing logic Archaeology.json 2`] = `[]`;

exports[`Snapshot of JSON parsing logic Archaeology.json 3`] = `
{
  "sh85006507": "Archaeology",
}
`;

exports[`Snapshot of JSON parsing logic History.json 1`] = `
[
  {
    "aL": [
      "Source material, Historical",
      "Primary sources (Historical sources)",
      "Historical sources",
      "Sources, Historical",
      "Historical source material",
    ],
    "lcc": "D5-D5.5",
    "note": "Here are entered collections of documents, records, etc. upon which narrative history is based.",
    "nt": [
      "sh2003003039",
      "sh85006913",
      "sh85022719",
      "sh85038194",
      "sh85135408",
    ],
    "pL": "History--Sources",
    "uri": "sh85061227",
  },
]
`;

exports[`Snapshot of JSON parsing logic History.json 2`] = `[]`;

exports[`Snapshot of JSON parsing logic History.json 3`] = `
{
  "sh85061227": "History--Sources",
}
`;

exports[`Snapshot of JSON parsing logic History-subdiv.json 1`] = `[]`;

exports[`Snapshot of JSON parsing logic History-subdiv.json 2`] = `
[
  {
    "aL": [
      "Frontier troubles",
    ],
    "note": "Use as a topical subdivision under names of countries, cities, etc., and individual corporate bodies, uniform titles of sacred works, and under classes of persons, ethnic groups, and topical headings.",
    "nt": [
      "sh00004047",
      "sh00005861",
      "sh00005863",
      "sh00005865",
      "sh00006030",
      "sh00006053",
      "sh00006054",
      "sh00006055",
      "sh2002012516",
      "sh2005000301",
      "sh2005003729",
      "sh2007008610",
      "sh2008009851",
      "sh2014001466",
      "sh2018001286",
      "sh85006507",
      "sh85006513",
      "sh85006887",
      "sh85007960",
      "sh85012434",
      "sh85014152",
      "sh85025619",
      "sh85031299",
      "sh85031326",
      "sh85033502",
      "sh85038179",
      "sh85038194",
      "sh85038418",
      "sh85050144",
      "sh85053742",
      "sh85061115",
      "sh85061208",
      "sh85064459",
      "sh85074517",
      "sh85077565",
      "sh85081954",
      "sh85085072",
      "sh85085207",
      "sh85088131",
      "sh85090384",
      "sh85100124",
      "sh85103033",
      "sh85107780",
      "sh85108672",
      "sh85113275",
      "sh85113507",
      "sh85114172",
      "sh85119320",
      "sh85120563",
      "sh85123948",
      "sh85124029",
      "sh85133488",
      "sh85148201",
      "sh88002668",
      "sh88004498",
      "sh89003666",
      "sh91002236",
      "sh98002432",
      "sh99004871",
      "sh99005023",
      "sh99005577",
      "sh99005617",
      "sh99005684",
      "sh99014676",
    ],
    "pL": "History",
    "uri": "sh99005024",
  },
]
`;

exports[`Snapshot of JSON parsing logic History-subdiv.json 3`] = `
{
  "sh99005024": "History",
}
`;

exports[`Snapshot of JSON parsing logic Obsidian.json 1`] = `
[
  {
    "aL": [
      "Hyalopsite",
      "Iceland agate",
    ],
    "bt": [
      "sh85144250",
    ],
    "lcc": "QE462.O28",
    "pL": "Obsidian",
    "uri": "sh85093752",
  },
]
`;

exports[`Snapshot of JSON parsing logic Obsidian.json 2`] = `[]`;

exports[`Snapshot of JSON parsing logic Obsidian.json 3`] = `
{
  "sh85093752": "Obsidian",
}
`;
